src/wasm/literal.cpp

```diff
@@ -1785,4 +1785,87 @@ Literal Literal::bitselectV128(const Literal& left,
   return andV128(left).orV128(notV128().andV128(right));
 }
 
+template<typename T> struct TwiceWidth {};
+template<> struct TwiceWidth<int8_t> { using type = int16_t; };
+template<> struct TwiceWidth<int16_t> { using type = int32_t; };
+
+template<typename T>
+Literal saturating_narrow(
+  typename TwiceWidth<typename std::make_signed<T>::type>::type val) {
```

Comment:
Does this handle when `val` is an unsigned number that exceeds its signed version's max?

Reply:
The narrowing operations always treat the input lanes as signed, so this can't happen.